### PR TITLE
Add container declarations to some projects

### PR DESCRIPTION
**Pull request by Sinerum**
Add containers declarations to bzip2, git, hypre and z3 and also add revision base compilation fixes to bzip2 and libxml2

### CODE REVIEW

File: varats/varats/projects/c_projects/bzip2.py [hunk 1]
```diff
@@ -91,20 +124,36 @@ def run_tests(self) -> None:
...
-
-        with local.cwd(bzip2_source / "build"):
-            with local.env(CC=str(cc_compiler), CXX=str(cxx_compiler)):
-                bb.watch(cmake)("..")
-
-            bb.watch(cmake)(
-                "--build", ".", "--config", "Release", "-j",
-                get_number_of_jobs(bb_cfg())
-            )
+        clang = bb.compiler.cc(self)
```
Comment (vulder) [inline varats/varats/projects/c_projects/bzip2.py]:
```suggestion
```

File: varats/varats/projects/c_projects/bzip2.py [hunk 2]
```diff
@@ -91,20 +124,36 @@ def run_tests(self) -> None:
...
+        if bzip2_version in typed_revision_range(
+            Bzip2._MAKE_VERSIONS, bzip2_version_source, ShortCommitHash
+        ):
+            with local.cwd(bzip2_source):
+                with local.env(CC=str(clang)):
```
Comment (vulder) [inline varats/varats/projects/c_projects/bzip2.py]:
```suggestion
                with local.env(CC=str(cc_compiler)):
```

File: varats/varats/projects/c_projects/bzip2.py [hunk 3]
```diff
@@ -91,20 +124,36 @@ def run_tests(self) -> None:
...
+                    bb.watch(make)("-j", get_number_of_jobs(bb_cfg()))
+        elif bzip2_version in typed_revision_range(
+            Bzip2._AUTOTOOLS_VERSIONS, bzip2_version_source, ShortCommitHash
+        ):
+            with local.cwd(bzip2_source):
+                with local.env(CC=str(clang)):
```
Comment (vulder) [inline varats/varats/projects/c_projects/bzip2.py]:
```suggestion
                with local.env(CC=str(cc_compiler)):
```

**Review by vulder — APPROVED** [review]
Looks good overall

File: varats/varats/projects/c_projects/bzip2.py [hunk 4]
```diff
@@ -82,7 +108,14 @@ def binaries_for_revision(
         binary_map = RevisionBinaryMap(get_local_project_git_path(Bzip2.NAME))
 
         binary_map.specify_binary('build/bzip2', BinaryType.EXECUTABLE)
-
+        binary_map.specify_binary(
```
Comment (boehmseb) [inline varats/varats/projects/c_projects/bzip2.py:117]: Is `build/bzip2` always valid or only in the revisions that are not in this range?

File: varats/varats/projects/c_projects/bzip2.py [hunk 5]
```diff
@@ -91,20 +124,36 @@ def run_tests(self) -> None:
     def compile(self) -> None:
         """Compile the project."""
         bzip2_source = local.path(self.source_of(self.primary_source))
```
Comment (boehmseb) [inline varats/varats/projects/c_projects/bzip2.py]:
```suggestion
        bzip2_source = local.path(self.source_of_primary)```

File: varats/varats/projects/c_projects/libxml2.py [hunk 6]
```diff
@@ -62,11 +63,17 @@ def run_tests(self) -> None:
     def compile(self) -> None:
         """Compile the project."""
         libxml2_version_source = local.path(self.source_of(self.primary_source))
```
Comment (boehmseb) [inline varats/varats/projects/c_projects/libxml2.py]:
```suggestion
        libxml2_version_source = local.path(self.source_of_primary)
```

**Review by boehmseb — CHANGES_REQUESTED** [review]

File: varats/varats/projects/c_projects/bzip2.py (hunk 4 above)
Comment (Sinerum) [reply varats/varats/projects/c_projects/bzip2.py:117]: Only for revisions not in the range which is equvivalent to revisions newer than e264a7f7c44fae62f5be9840946f6bc0e8cd6512

Comment (boehmseb) [reply varats/varats/projects/c_projects/bzip2.py:117]: Then this should also be reflected in the map.

File: varats/varats/projects/c_projects/bzip2.py [hunk 7]
```diff
@@ -81,30 +107,58 @@ def binaries_for_revision(
...
         return binary_map[revision]
 
     def run_tests(self) -> None:
         pass
 
     def compile(self) -> None:
         """Compile the project."""
-        bzip2_source = local.path(self.source_of(self.primary_source))
-
+        bzip2_source = local.path(self.source_of_primary)
```
Comment (boehmseb) [inline varats/varats/projects/c_projects/bzip2.py]: It should be possible to use the `pathlib.Path` object in all locations where you use this.
At the very least, it should be possible to do `str(bzip2_version_source)`.
Could you please try? Then we could get rid of this duplication of variables that plumbum always causes here.

**Review by boehmseb — APPROVED** [review]

**Pull request merged**